File: services/scriptmgr/scripts/exec.py
#!/usr/bin/python -W ignore::DeprecationWarning

import  sys
import  os
import  socket
import  signal
import  string
import  time
import  resource
import  urllib2, urllib
import  optparse
import  scraperwiki

try    : import json
except : import simplejson as json


    # Unfortunately necessary to do this because PYTHONUNBUFFERED=True does nto get good enough results and tends to still concatenate lines when short and rapid
class ConsoleStream:
    def __init__(self, fd):
        self.m_text = ''
        self.m_fd = fd

    def saveunicode(self, text):
        try:    return unicode(text)
        except UnicodeDecodeError:     pass
        try:    return unicode(text, encoding='utf8')
        except UnicodeDecodeError:     pass
        try:    return unicode(text, encoding='latin1')
        except UnicodeDecodeError:     pass
        return unicode(text, errors='replace')
    
    def write(self, text):
        self.m_text += self.saveunicode(text)
        if self.m_text and self.m_text[-1] == '\n' :
            self.flush()

    def flush(self) :
        if self.m_text:
            scraperwiki.dumpMessage({'message_type': 'console', 'content': self.m_text})
            self.m_text = ''
            self.m_fd.flush()
            
    def close(self):

        self.m_fd.close()

    def fileno(self):
        return self.m_fd.fileno()


# Make sure we have unbuffered outputs
#sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)
#sys.stderr = os.fdopen(sys.stderr.fileno(), 'w', 0)

scraperwiki.logfd = sys.stderr
sys.stdout = ConsoleStream(scraperwiki.logfd)
sys.stderr = ConsoleStream(scraperwiki.logfd)

parser = optparse.OptionParser()
parser.add_option("--script", metavar="name")    # not the scraper name, this is tmp file name which we load and execute
parser.add_option("--ds", metavar="server:port")
parser.add_option("--gid")    # nogroup
parser.add_option("--uid")    # nobody
parser.add_option("--scrapername")
parser.add_option("--runid")
parser.add_option("--qs")
parser.add_option("--path")
options, args = parser.parse_args()

if options.gid:
    os.setregid(int(options.gid), int(options.gid))
if options.uid:
    os.setreuid(int(options.uid), int(options.uid))
if options.path:
    sys.path.append( options.path )
if options.qs:
    os.environ['QUERY_STRING'] = options.qs
    os.environ['URLQUERY'] = options.qs    


host, port = string.split(options.ds, ':')

# Added two new arguments as this seems to have changed in scraperlibs
scraperwiki.datastore.create(host, port, options.scrapername or "", options.runid, [], 5000)



resource.setrlimit(resource.RLIMIT_CPU, (80, 82,))

#  Set up a CPU time limit handler which simply throws a python so it can be handled cleanly before the hard limit is reached
def sigXCPU(signum, frame) :
    raise Exception("ScraperWiki CPU time exceeded")
signal.signal(signal.SIGXCPU, sigXCPU)

code = open(options.script).read()
try:
    import imp
    mod = imp.new_module('scraper')
    exec code.rstrip() + "\n" in mod.__dict__
except Exception, e:
    etb = scraperwiki.stacktrace.getExceptionTraceback(code)  
    assert etb.get('message_type') == 'exception'
    scraperwiki.dumpMessage(etb)
except SystemExit, se:
    sys.stdout.flush()
    sys.stderr.flush()

    # If we do not temporarily yield a slice of the CPU here then the launching 
    # process will not be able to read from stderr before we exit.
    import time 
    time.sleep(0)

    raise se

sys.stdout.flush()
sys.stderr.flush()

File: services/scriptmgr/scripts/exec.rb
#!/usr/bin/ruby

$stdout.sync = true

require 'rubygems'   # for nokigiri to work on all machines, and for JSON/Iconv on OSX
require 'json'
require 'date'
require 'iconv'
require 'optparse'
require	'scraperwiki'
require 'scraperwiki/datastore'
require 'scraperwiki/stacktrace'

$logfd  = $stderr

class ConsoleStream
    def initialize(fd)
        @fd   = fd
        @text = ''
    end

    # Do our best to turn anything into unicode, for display on console
    # (for datastore, we give errors if it isn't already UTF-8)
    def saveunicode(text)
        begin
            text = Iconv.conv('utf-8', 'utf-8', text)
        rescue Iconv::IllegalSequence
            begin
                text = Iconv.conv('utf-8', 'iso-8859-1', text)
            rescue Iconv::IllegalSequence
                text = Iconv.conv('utf-8//IGNORE', 'utf-8', text)
            end
        end
        return text
    end

    def write(text)
        @text = @text + saveunicode(text)
        if @text.length > 0 && @text[-1] == "\n"[0]
            flush
        end
    end

    def <<(text)
       write text
    end

    def flush
      if @text != ''
          ScraperWiki.dumpMessage( { 'message_type' => 'console', 'content' => @text }  )
          @fd.flush
          @text = ''
      end
    end

    def close
        @fd.close()
    end
end

$stdout = ConsoleStream.new($logfd)
$stderr = ConsoleStream.new($logfd)

Process.setrlimit(Process::RLIMIT_CPU, 80, 82) 
 
Signal.trap("XCPU") do
    raise Exception, "ScraperWiki CPU time exceeded"
end

options = {}
OptionParser.new do|opts|
   opts.on( '--script=[SCRIPT]') do|script|
     options[:script] = script
   end
   opts.on( '--ds=[DS]') do|ds|
     options[:ds] = ds
   end
   opts.on( '--gid=[GID]') do|gid|
     Process::Sys.setregid(gid.to_i, gid.to_i)
   end
   opts.on( '--uid=[UID]') do|uid|
     Process::Sys.setreuid(uid.to_i, uid.to_i)
   end
   opts.on( '--scrapername=[SCRAPERNAME]') do|scrapername|
     options[:scrapername] = scrapername
   end
   opts.on( '--qs=[QSTRING]') do|qstring|
     ENV['QUERY_STRING'] = qstring
     ENV['URLQUERY'] = qstring
   end   
   opts.on( '--runid=[RUNID]') do|runid|
     options[:runid] = runid
   end
end.parse(ARGV)

host, port = options[:ds].split(':')
SW_DataStore.create(host, port, options[:scrapername], options[:runid])

code = File.new(options[:script], 'r').read()
begin
    #eval code # this doesn't give you line number of top level errors, instead we use require_relative:
    require_relative options[:script]
rescue Exception => e
    est = getExceptionTraceback(e, code, options[:script])
    # for debugging:
    # File.open("/tmp/fairuby", 'a') {|f| f.write(JSON.generate(est)) }
    ScraperWiki.dumpMessage(est)
end

$stdout.flush
$stderr.flush
